### DISCUSSION TITLE
Preload neighbourind levels with LevelSet

### DISCUSSION THREAD

**Discussion by Ggcpp** [Trouv/bevy_ecs_ldtk#316] [Q&A]
Hi,

I have custom spawing behaviors for my levels but I still want to preload neighbouring levels around the player's current level, how can I do that with ```LevelSet```? I saw that there is a ```LoadedLevel::neighbours(&self)``` function but I don't know how to retrieve a ```LoadedLevel``` from the current state of the game, maybe it's the wrong approach?

I'm very new to all this stuff, sorry for all my questions :)

**Comment by Trouv** [comment]
No worries, happy to help.

So, this is one problem where you will need to access the asset data. You don't necessarily need a loaded level though, an `ldtk::Level` will have the info you need too. A `LoadedLevel` is simply a wrapper around a `Level` that has type guarantees that the internal level data is complete. For more info: https://docs.rs/bevy_ecs_ldtk/latest/bevy_ecs_ldtk/assets/struct.LdtkProject.html#raw-vs-loaded-levels 

Firstly, you need access to the raw `LdtkProject` from the asset store. For this you need to parameterize your system with `ldtk_project_assets: Res<Assets<LdtkProject>>` and query it with a `Handle<LdtkProject>`. How exactly you retrieve the handle depends on how you're handling it in your game. If you only have one project spawned you can just parameterize `ldtk_projects: Query<Handle<LdtkProject>>` then grab it with `let project_handle = ldtk_projects.single()`. Then use that to retrieve the asset data with `ldtk_project_assets.get(project_handle).unwrap();`.

Once you have the raw project data there are actually many methods you can use to get the data for a particular level on the following traits:
https://docs.rs/bevy_ecs_ldtk/latest/bevy_ecs_ldtk/ldtk/raw_level_accessor/trait.RawLevelAccessor.html
https://docs.rs/bevy_ecs_ldtk/latest/bevy_ecs_ldtk/assets/trait.LevelMetadataAccessor.html
I'm guessing you'll want to use [`LdtkProject::get_raw_level_by_iid`](https://docs.rs/bevy_ecs_ldtk/latest/bevy_ecs_ldtk/assets/trait.LevelMetadataAccessor.html#method.get_raw_level_by_iid).

Then the `Level` has a `neighbors` field you can iterate through like so:
```rust
let neighboring_levels = level.neighbors.iter().map(|n| LevelIid::new(n.level_iid.clone()));
```

This guide in the book isn't exactly what you want to do, but it does go through the process of getting level data, which might be helpful:
https://trouv.github.io/bevy_ecs_ldtk/latest/how-to-guides/make-level-selection-follow-player.html#determine-bounds-of-spawned-levels-and-update-level-selection
